File: packages/typescript/core-sdk/src/auth.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import http from "node:http";
import { PasswordTokenSource, StaticTokenSource } from "./auth.js";

async function withServer(handler: http.RequestListener, run: (baseUrl: string) => Promise<void>) {
  const server = http.createServer(handler);
  await new Promise<void>((resolve) => server.listen(0, resolve));
  const address = server.address();
  const port = typeof address === "object" && address ? address.port : 0;
  try {
    await run(`http://127.0.0.1:${port}`);
  } finally {
    await new Promise<void>((resolve) => server.close(() => resolve()));
  }
}

test("PasswordTokenSource mints and caches a token", async () => {
  let calls = 0;
  await withServer(
    (req, res) => {
      calls++;
      let raw = "";
      req.on("data", (chunk) => (raw += chunk));
      req.on("end", () => {
        const form = new URLSearchParams(raw);
        assert.equal(form.get("grant_type"), "password");
        assert.equal(form.get("username"), "demo");
        res.writeHead(200, { "Content-Type": "application/json" }).end(
          JSON.stringify({ access_token: "minted-token", expires_in: 300 }),
        );
      });
    },
    async (baseUrl) => {
      const ts = new PasswordTokenSource({
        keycloakUrl: baseUrl,
        realm: "core",
        clientId: "core-platform",
        username: "demo",
        password: "demo",
      });
      const first = await ts.token();
      assert.equal(first, "minted-token");
      const second = await ts.token();
      assert.equal(second, "minted-token");
      assert.equal(calls, 1, "second call within the token lifetime should hit the cache, not mint again");
    },
  );
});

test("PasswordTokenSource refreshes once the cached token is near expiry", async () => {
  let calls = 0;
  await withServer(
    (_req, res) => {
      calls++;
      res.writeHead(200, { "Content-Type": "application/json" }).end(
        JSON.stringify({ access_token: `token-${calls}`, expires_in: 1 }),
      );
    },
    async (baseUrl) => {
      const ts = new PasswordTokenSource({
        keycloakUrl: baseUrl,
        realm: "core",
        clientId: "core-platform",
        username: "demo",
        password: "demo",
        refreshSkewMs: 5000,
      });
      await ts.token();
      await ts.token();
      assert.equal(calls, 2, "the second call should trigger a real refresh (token expires inside the skew window)");
    },
  );
});

test("StaticTokenSource never calls out", async () => {
  const ts = new StaticTokenSource("fixed");
  assert.equal(await ts.token(), "fixed");
});

File: packages/flutter/core_sdk/test/auth_test.dart
import 'package:test/test.dart';
import 'package:core_sdk/core_sdk.dart';

import 'test_server.dart';

void main() {
  test('PasswordTokenSource mints and caches a token', () async {
    var calls = 0;
    final server = await TestServer.start((req) async {
      calls++;
      final body = await readBody(req);
      final form = Uri.splitQueryString(body);
      expect(form['grant_type'], 'password');
      expect(form['username'], 'demo');
      await respondJson(req, 200, {'access_token': 'minted-token', 'expires_in': 300});
    });
    addTearDown(server.close);

    final ts = PasswordTokenSource(
      keycloakUrl: server.baseUrl,
      realm: 'core',
      clientId: 'core-platform',
      username: 'demo',
      password: 'demo',
    );

    final first = await ts.token();
    expect(first, 'minted-token');
    final second = await ts.token();
    expect(second, 'minted-token');
    expect(calls, 1, reason: 'second call within the token lifetime should hit the cache, not mint again');
  });

  test('PasswordTokenSource refreshes once the cached token is near expiry', () async {
    var calls = 0;
    final server = await TestServer.start((req) async {
      calls++;
      await respondJson(req, 200, {'access_token': 'token-$calls', 'expires_in': 1});
    });
    addTearDown(server.close);

    final ts = PasswordTokenSource(
      keycloakUrl: server.baseUrl,
      realm: 'core',
      clientId: 'core-platform',
      username: 'demo',
      password: 'demo',
      refreshSkew: const Duration(seconds: 5),
    );

    await ts.token();
    await ts.token();
    expect(calls, 2, reason: 'the second call should trigger a real refresh (token expires inside the skew window)');
  });

  test('StaticTokenSource never calls out', () async {
    final ts = StaticTokenSource('fixed');
    expect(await ts.token(), 'fixed');
  });
}
